### DISCUSSION TITLE
A question about self define kernel

### DISCUSSION THREAD

**Discussion by wangzhaohao** [idaholab/moose#21587] [Q&A General] [answered]
Hi, everyone! I want to define a ```PDE``` which form is 
$\bigtriangledown \bigtriangledown C +\bigtriangledown (\frac{640C}{T^{2}}\bigtriangledown T) =0$
and the ```bcs``` is
$\bigtriangledown C = 0   (x =0)$
$C =1.6 (x=1)$
I define three kernels ConcentrationConvectionSource ConcentrationConvectionTemp ConcentrationDiffusionTemp
and a .i file
concentration_back.i
they are all follow.
https://github.com/wangzhaohao/termite2/tree/main/termite2
![微信图片_20220714211733](https://user-images.githubusercontent.com/51054734/178991630-dc6c3d92-a503-4628-997e-cca9c70cecc9.png)
![微信图片_20220714211752](https://user-images.githubusercontent.com/51054734/178991646-025add77-9347-402d-9d59-f135ca148496.png)

but the result is different with Finite Difference, I meet this problem so long,but i can't find the wrong, so who can give me some advice, Thanks for your help!
hao

**Comment by GiudGiud** [comment, marked as answer]
Hello

I took a quick look. In concentration.i you set DirichletBC instead of a NeumannBC on the left boundary
```
  [left]
    type = DirichletBC
    variable = concentration
    boundary = left
    value = 0
  []
```
this isnt the result shown in your post either.

Instead of sharing a zip of your folder, please create a github repo, share the link here and tell us which inputs should reproduce the problem here. 

Also is temperature just fixed?

Guillaume

**Reply by wangzhaohao** [reply]
the.i file is concentration_back.i, not the concontration.i

**Reply by GiudGiud** [reply]
that input has kernels for an auxvariable (temp), which shouldnt even be allowed, and doesnt converge well.

**Reply by wangzhaohao** [reply]
https://github.com/wangzhaohao/termite2/tree/main/termite2
this is my code on github. the auxkernel is used for i just want to give a 
T = -1500*x*x+2000
I follow the Thermal Expansion, which give temp (also a auxvariable) by auxkernel.

**Comment by WilkAndy** [comment]
Have you got one equation and two unknowns (C and T)?

**Reply by wangzhaohao** [reply]
the variable T is give as auxvariable.

**Comment by YaqiWang** [comment]
Is the PJFNK solve converged? What makes you think you FD solution is accurate? How big is the difference between FD and your FEM solution?

**Reply by wangzhaohao** [reply]
there is no warning. and i think the FD solution is accurate is that when i set the bcs
```
x = 0, C = 0
x = 1, C = 1.6
```
![c77f12cd9f378325204642b033252cf](https://user-images.githubusercontent.com/51054734/179229353-727c838c-044f-4706-8921-805ef270c6fd.jpg)

![8aff83b4e91c73f069fa5323637035d](https://user-images.githubusercontent.com/51054734/179229407-ed948787-e9b3-4d32-a133-e485bd9bfa23.jpg)


there have analytical solution. the FD solution satisfy, but the FEM is not.  so i guess there is something wrong about myself define kernels.

**Reply by YaqiWang** [reply]
This does not make sense. If your temperature is constant, its gradient is zero and all the three kernels coupling the temperature have zero contributions to the residual, so only the diffusion kernel and two Dirichlet bcs left, you should get a linear solution from 0 to 1.6.

**Reply by GiudGiud** [reply]
Temperature is constant in time, it has a spatial profile

**Reply by wangzhaohao** [reply]
> 
so i think there is somthing which i don't  notice in my self define kernel

**Reply by GiudGiud** [reply]
to debug you will want to take it step by step. 

The problem is easy enough that you can replace each kernel one by one to make sure they are correct. When you reproduce the analytical solution for a simplified problem, you can consider the objects currently in use are correct, and investigate the others. 

Another option is to replace your kernels with framework ones, which are already correct. 
For example you may use ADMatDiffusion 
https://mooseframework.inl.gov/source/kernels/ADMatDiffusion.html
with a ParsedMaterialProperty (use the AD version) for your diffusion term
https://mooseframework.inl.gov/source/materials/ParsedMaterial.html

In any case you will want to remove the kernels acting on an auxiliary variable
